typed_sql/doc/07_aggregate_functions.md

```diff
@@ -0,0 +1,339 @@
+query, as illustrated below:
+
+```sql
+SELECT authorId, SUM(stock) FROM books GROUP BY authorId;
+```
+
+In Dart we can do the same thing using `.groupBy` and `.aggregate` as follows:
+```dart bookstore_test.dart#sum-books-group-by-author
+final result = await db.books
+    .groupBy((b) => (b.author,))
+    .aggregate(
+      (agg) => agg
+          // aggregates:
+          .sum((book) => book.stock)
+          .count(),
+    )
+    .select(
+      (author, stock, countBooksByAuthor) => (
+        author.name,
+        stock,
+        countBooksByAuthor,
+      ),
+    )
+    .fetch();
+
+check(result).unorderedEquals([
+  // Author, total stock, count of books
+  ('Easter Bunny', 22, 3),
+  ('Bucks Bunny', 45, 2),
+]);
+```
+
+The `.groupBy` method is used to make a projection (like `.select`), each
+distinct row of the projection becomes a group, and rows in said group is
+aggregated using `.aggregate`.
+
+The `.aggregate` function give you an _aggregation builder_ `agg` which is used
+to build aggregations. By adding each aggregate function with chained method
+calls the `.aggregate` function can return query that is the combination of
+projection from the `.groupBy` and the aggregations.
+
+The following `.select` is entirely optional, without the result would have been
+`List<(Author, int, int)>` instead of `List<(String, int, int)>`.
+
+> [!TIP]  
+> The optional `// aggregates:` comment after `(agg) => agg` forces the
+> Dart formatter put each aggregated column on a new line.
+
+The astute reader might have noticed that the SQL query given at the start of
+this section does in fact not return the _author name_, instead it's grouped by
+`authorId`, and would return the `authorId`. This is because the `b.author` in
+the `.groupBy` projection is actually a reference, which in turn makes the
+`author.name` in the final `.select` projection a subquery. We could avoided
+this by simply using `b.authorId` in the `.groupBy` projection, as illustrated
+below:
+```dart bookstore_test.dart#sum-books-group-by-authorId
+final result = await db.books
+    .groupBy((b) => (b.authorId,))
+    .aggregate(
+      (agg) => agg
+          // aggregates:
+          .sum((book) => book.stock)
+          .count(),
+    )
+    .select(
+      (authorId, stock, countBooksByAuthor) => (
+        authorId,
+        stock,
+        countBooksByAuthor,
+      ),
+    )
+    .fetch();
+
+check(result).unorderedEquals([
+  // AuthorId, total stock, count of books
+  (1, 22, 3),
+  (2, 45, 2),
+]);
+```
+
+But in practice, using references like `b.author` instead of `b.authorId` make
+it easy to lookup associated rows in the database. However, for complex queries
+subqueries can sometimes be slow. The next section on `.join` illustrates how
+to effectively avoid subqueries. Though it can be wise to not optimize
+prematurely, and instead prefer to optimize slow queries when they become
+a problem.
```

Comment:
Do we really want to advise users when to optimize?

Reply:
maybe not,